#### What we will learn:
- Read data file by pandas to dataframe
- Set column names to dataframe
- Investigate data
- Filter data in columns
- Visualization missing values by histogram
- Drop null data
- Save file to CSV without adding row index

**Dataset** <br>
[Adult dataset from UCI](https://archive.ics.uci.edu/ml/datasets/Adult "dataset")


In [1]:
import pandas as pd

# Set ipython's max row display
pd.set_option('display.max_row', 1000)
# Set iPython's max column width to 50
pd.set_option('display.max_columns', 50)

# Read data file
# delimiter=', '           set separator 
# names = ['', '', ..],    set column names
df = pd.read_csv("adult.data", header=None, delimiter=', ', 
                 names = ['age', 'workclass', 'fnlwgt', 
                  'education', 'education-num', 'marital-status',
                  'occupation', 'relationship', 'race', 'sex', 
                  'capital-gain', 'capital-loss','hours-per-week','nativecountry', 'income'])

# Another way to add column
# df.columns = ['age', 'workclass', 'fnlwgt', 
#               'education', 'education-num', 'marital-status',
#               'occupation', 'relationship', 'race', 'sex', 
#               'capital-gain', 'capital-loss','hours-per-week','nativecountry', 'income']

# Display the first 10 rows and confirm is the column names and their values are correct.
df.head(10)

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:15: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  from ipykernel import kernelapp as app


age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   
5   37           Private  284582    Masters             14   
6   49           Private  160187        9th              5   
7   52  Self-emp-not-inc  209642    HS-grad              9   
8   31           Private   45781    Masters             14   
9   42           Private  159449  Bachelors             13   

          marital-status         occupation   relationship   race     sex  \
0          Never-married       Adm-clerical  Not-in-family  White    Male   
1     Married-civ-spouse    Exec-managerial        Husband  White    Male   
2               Divorced  Handlers-cleaners  Not-in-family  White    Male   
3     Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4     Married-civ-spouse     Prof-specialty           Wife  Black  Female   
5     Married-civ-spouse    Exec-managerial           Wife  White  Female   
6  Married-spouse-absent      Other-service  Not-in-family  Black  Female   
7     Married-civ-spouse    Exec-managerial        Husband  White    Male   
8          Never-married     Prof-specialty  Not-in-family  White  Female   
9     Married-civ-spouse    Exec-managerial        Husband  White    Male   

   capital-gain  capital-loss  hours-per-week  nativecountry income  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K  
5             0             0              40  United-States  <=50K  
6             0             0              16        Jamaica  <=50K  
7             0             0              45  United-States   >50K  
8         14084             0              50  United-States   >50K  
9          5178             0              40  United-States   >50K

In [2]:
#Show number of rows and columns in format of (row, column)
df.shape

(32561, 15)

In [3]:
#Show column information 
# e.g. column name, number of rows data, null or non-null, data type (object means string)
# At this step we should confirm the data type and missing value.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
age               32561 non-null int64
workclass         32561 non-null object
fnlwgt            32561 non-null int64
education         32561 non-null object
education-num     32561 non-null int64
marital-status    32561 non-null object
occupation        32561 non-null object
relationship      32561 non-null object
race              32561 non-null object
sex               32561 non-null object
capital-gain      32561 non-null int64
capital-loss      32561 non-null int64
hours-per-week    32561 non-null int64
nativecountry     32561 non-null object
income            32561 non-null object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [4]:
# Another way to get number of columns that contain null.
df.isnull().any().sum()

0

In [5]:
# statistical check for numerical data
df.describe()

age        fnlwgt  education-num  capital-gain  capital-loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours-per-week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

**Note: We can't see missing value here but the data has really been processed and ready for next step?**

Let's check values in each column

In [6]:
# Loop to print unique value and amount for each column name where the data type is 'object'.
# 
def print_unique_vals(df):
    for d in df.columns:
        if df[d].dtypes == object:
            print("Column name:", d)
        
            print(df[d].value_counts())
        
print_unique_vals(df)      

Column name: workclass
Private             22696
Self-emp-not-inc     2541
Local-gov            2093
?                    1836
State-gov            1298
Self-emp-inc         1116
Federal-gov           960
Without-pay            14
Never-worked            7
Name: workclass, dtype: int64
Column name: education
HS-grad         10501
Some-college     7291
Bachelors        5355
Masters          1723
Assoc-voc        1382
11th             1175
Assoc-acdm       1067
10th              933
7th-8th           646
Prof-school       576
9th               514
12th              433
Doctorate         413
5th-6th           333
1st-4th           168
Preschool          51
Name: education, dtype: int64
Column name: marital-status
Married-civ-spouse       14976
Never-married            10683
Divorced                  4443
Separated                 1025
Widowed                    993
Married-spouse-absent      418
Married-AF-spouse           23
Name: marital-status, dtype: int64
Column name: occupation
Prof

**Alert! columns workclass and occupation contain '?'; we need to pre-process this dataset**

In [7]:
# Check number of row data that contain "?" from 'workclass' and 'occupation'
df_temp = df.loc[(df['workclass'] == '?') & (df['occupation'] == '?')]
print("Number of rows contain '?' in column workclass and occupation: ", df_temp.shape[0])
df_temp.head()

Number of rows contain '?' in column workclass and occupation:  1836


age workclass  fnlwgt     education  education-num  \
27    54         ?  180211  Some-college             10   
61    32         ?  293936       7th-8th              4   
69    25         ?  200681  Some-college             10   
77    67         ?  212759          10th              6   
106   17         ?  304873          10th              6   

            marital-status occupation   relationship                race  \
27      Married-civ-spouse          ?        Husband  Asian-Pac-Islander   
61   Married-spouse-absent          ?  Not-in-family               White   
69           Never-married          ?      Own-child               White   
77      Married-civ-spouse          ?        Husband               White   
106          Never-married          ?      Own-child               White   

        sex  capital-gain  capital-loss  hours-per-week  nativecountry income  
27     Male             0             0              60          South   >50K  
61     Male             0             0              40              ?  <=50K  
69     Male             0             0              40  United-States  <=50K  
77     Male             0             0               2  United-States  <=50K  
106  Female         34095             0              32  United-States  <=50K

**Let's investigate samples with missing values**

In [8]:
print_unique_vals(df_temp)

Column name: workclass
?    1836
Name: workclass, dtype: int64
Column name: education
HS-grad         532
Some-college    514
Bachelors       173
11th            118
10th            100
7th-8th          72
Assoc-voc        61
9th              51
Masters          48
Assoc-acdm       47
12th             40
5th-6th          30
Prof-school      18
Doctorate        15
1st-4th          12
Preschool         5
Name: education, dtype: int64
Column name: marital-status
Never-married            766
Married-civ-spouse       636
Divorced                 184
Widowed                  153
Separated                 66
Married-spouse-absent     29
Married-AF-spouse          2
Name: marital-status, dtype: int64
Column name: occupation
?    1836
Name: occupation, dtype: int64
Column name: relationship
Own-child         538
Husband           489
Not-in-family     439
Unmarried         175
Wife              132
Other-relative     63
Name: relationship, dtype: int64
Column name: race
White                 15

In [9]:
import matplotlib.pyplot as plt

# Loop to show histogram in missing value samples
for d in df_temp.columns:
    if df_temp[d].dtypes == object:
        fig, ax = plt.subplots()
        df_temp[d].value_counts().plot(kind='barh', title=d)


**We have 32561 rows in total while 1836 rows contain '?'** <br>
If we drop those data, we still have large number. So, let's try dropping.

In [10]:
df.dropna(inplace=True)
print(df.shape)
df.head()

(32561, 15)


age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week  nativecountry income  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

In [11]:
df.to_csv("adult_drop_missing.csv", index=False)

In [12]:
df_new = pd.read_csv("adult_drop_missing.csv")
df_new.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week  nativecountry income  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K